fuel-block-producer/src/block_producer/transaction_selector.rs

```diff
@@ -0,0 +1,180 @@
+                )
+                .gas_price(tx_gas.price)
+                .gas_limit(tx_gas.limit)
+                .add_unsigned_coin_input(
+                    rng.gen(),
+                    rng.gen(),
+                    1_000_000,
+                    Default::default(),
+                    Default::default(),
+                    0,
+                )
+                .add_output(Output::Change {
+                    to: Default::default(),
+                    amount: 0,
+                    asset_id: Default::default(),
+                })
+                // The block producer assumes transactions are already checked
+                // so it doesn't need to compute valid sigs for tests
+                .finalize_checked_without_signature(
+                    0,
+                    &ConsensusParameters {
+                        gas_price_factor: 1,
+                        ..ConsensusParameters::default()
+                    },
+                )
+            })
+            .map(Arc::new)
+            .collect();
+
+        select_transactions(
+            txs,
+            &Config {
+                max_gas_per_block: block_gas_limit,
+                ..Default::default()
+            },
+        )
+        .into_iter()
+        .map(|tx| {
+            let tx = tx.transaction();
+            TxGas {
+                limit: tx.gas_limit(),
+                price: tx.gas_price(),
+            }
+        })
+        .collect()
+    }
+
+    #[test]
+    fn selector_works_with_empty_input() {
+        let selected = make_txs_and_select(&[], 1_000_000);
+        assert!(selected.is_empty());
+    }
+
+    #[test]
+    fn selector_prefers_highest_gas_txs_and_sorts() {
+        #[rustfmt::skip]
+        let original = [
+            TxGas { price: 3, limit: 2000 },
+            TxGas { price: 1, limit: 1000 },
+            TxGas { price: 4, limit: 3000 },
+            TxGas { price: 5, limit: 1000 },
+            TxGas { price: 2, limit: 1000 },
+        ];
+
+        #[rustfmt::skip]
+        let ordered = [
+            TxGas { price: 5, limit: 1000 },
+            TxGas { price: 4, limit: 3000 },
+            TxGas { price: 3, limit: 2000 },
+            TxGas { price: 2, limit: 1000 },
+            TxGas { price: 1, limit: 1000 },
+        ];
+
+        let selected = make_txs_and_select(&original, 1_000_000);
+        assert_eq!(ordered, selected.as_slice());
+    }
+
+    #[test]
+    fn selector_doesnt_exceed_max_gas_per_block() {
+        #[rustfmt::skip]
+        let original = [
+            TxGas { price: 3, limit: 2000 },
+            TxGas { price: 1, limit: 1000 },
+            TxGas { price: 4, limit: 3000 },
+            TxGas { price: 5, limit: 1000 },
+            TxGas { price: 2, limit: 1000 },
+        ];
+
+        for k in 0..original.len() {
+            for perm in original.into_iter().permutations(k) {
+                for gas_limit in [999, 1000, 2000, 2500, 3000, 5000, 6000, 10_000] {
```

Comment:
Maybe checking the number of transactions also is a good idea.
`[(999, 0), (1000, 1), (2000, 2), (2500, 2), (3000, 1), (5000, 3), (6000, 3), (10_000, 5)]`

Reply:
I think this one is less important, since the number of txs doesn't really matter as much as verifying we can never select more than the allowed amount of gas. It's also not that straightforward to configure the expected number of txs since we are performing these selections across all permutations, which in some cases means we have a set size of zero to select from for example.

Reply:
I'll look into adding the selection correctness to the preceding test: `selector_prefers_highest_gas_txs_and_sorts`

Reply:
Making these tests revealed a flaw in the transaction selection which I've now addressed!

The selection was using `tx.max_fee()` to calculate the block capacity, when it should've been using only gas instead (block capacity isn't dependent on gas price). This required a small change to fuel-tx, and so the patch section is just temporary until https://github.com/FuelLabs/fuel-tx/pull/189 gets merged.